src/drivers/virtio/virtqueue/packed.rs

```diff
@@ -963,17 +989,25 @@ impl Virtq for PackedVq {
 		// Zero transfers are not allowed
 		assert!(!tkns.is_empty());
 
-		let (next_off, next_wrap) = self.descr_ring.borrow_mut().push_batch(tkns);
+		let next_idx = self.descr_ring.borrow_mut().push_batch(tkns);
 
 		if notif {
-			self.drv_event
-				.borrow_mut()
-				.enable_specific(next_off, next_wrap);
+			self.drv_event.borrow_mut().enable_specific(next_idx);
 		}
 
-		if self.dev_event.is_notif() | self.dev_event.is_notif_specfic(next_off, next_wrap) {
-			self.notif_ctrl
-				.notify_dev(self.index.0, next_off, next_wrap);
+		let range = self.last_notify.get()..next_idx;
+		let notif_specfic = self
+			.dev_event
+			.notif_specfic()
+			.map_or(false, |idx| range.wrapping_contains(&idx));
+		self.last_notify.set(next_idx);
```

Comment:
Good question!

I think we should only notify for any transfers just performed. I don't think a device would except a notification for an old transfer. That would also potentially create problems with wrapping around.

What do you think?

Reply:
I think that would depend on the device checking to see if the specific descriptor it wants to request a notification for has already been written. If it does not, it would not be aware of the fact that it is an old dispatch.

What would be the problem with wrap-arounds? The driver cannot wrap around more than once before sending a notification to the device since the slots on the ring would be exhausted, if that is relevant.

Reply:
Good point. I changed it. 👍